mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,449 @@
+            return true;
+        }
+    }
+
+    /**
+     * Removes a member from the clan.
+     * @param transaction The tx
+     * @param player player to remove from the clan
+     * @return return true if it went ok, false if member is not part of clan or already in a clan.
+     * @throws IllegalArgumentException if member tried to remove was the leader of the clan
+     */
+    public boolean removeClanMember(Transaction transaction, SolarPlayer player) {
+        ClanDataObject receiver = player.getData(ClansKey.INSTANCE);
+        if (receiver.isSimilar(leader))  {
+            throw new IllegalArgumentException("Tried to remove leader of clan!");
+        }
+
+        int res = transaction.getProperty(DSLContext.class)
+                .deleteFrom(CLANS_CLAN_MEMBERSHIP)
+                .where(CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(this.clanID).and(CLANS_CLAN_MEMBERSHIP.USER_ID.eq(receiver.getUserId())))
+                .execute();
+
+        if (res != 1)  {
+            return false;
+        } else {
+            receiver.setCachedClan(null);
+            return true;
+        }
+    }
+
+    /**
+     * Adds clan as an ally
+     *
+     * @param transaction the tx
+     * @param receiver the receiver
+     * @throws IllegalArgumentException if receiver is already allied with this clan or another clan, or if sender already has an ally.
+     * @return true if ok, false if clans are already allied or already have separate allies in the table.
+     */
+    public boolean addClanAsAlly(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan ally as same clan!");
+
+        if (this.getAlliedClan(transaction).isPresent()) throw new IllegalArgumentException("Sender already have an ally!");
+        if (receiver.getAlliedClan(transaction).isPresent()) throw new IllegalArgumentException("Selected clan already has an ally!");
+
+        //note to aurium - these are now okay to do becaue we run the checks previous
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ALLIANCES)
+                .columns(CLANS_CLAN_ALLIANCES.CLAN_ID,CLANS_CLAN_ALLIANCES.ALLY_ID)
+                .values(this.clanID,receiver.getID())
+                .values(receiver.getID(),this.clanID)
+                .onDuplicateKeyIgnore()
+                .execute();
+
+        if (res != 2) {
+            return false;
+        } else {
+            manager.insertAllianceCache(this.getID(),receiver.getID());
+            return true;
+        }
+    }
+
+    /**
+     * Revokes ally. Can be called from either clan. Revokes alliance for both.
+     * @param transaction the tx
+     * @return true if ok, false if one of the clans or both of the clans has no ally
+     * @throws IllegalArgumentException if the clan does not have an ally
+     */
+    public boolean revokeAlly(Transaction transaction) {
```

Comment:
you need to take another look at this method too, both its specification and implementation.